tools/generate_property_translations.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+#
+# Generates translation strings from gecko property files for lupdate
+# to process
+
+import os, os.path, sys, re, dataclasses, functools, argparse
+
+DEFAULT_PREFIX = "sailfish_components_webview_popups-la"
+
+class Generator:
+    RE = re.compile(r"(?P<name>[\w-]+)\s*=\s*(?P<text>.*)\s*")
+    SPLITTER = re.compile(r"([A-Za-z0-9][a-z0-9]*)")
+    PLACEHOLDER = re.compile(r"%(?:([0-9]+)\$)?S")
+
+    def __init__(self, file):
+        self.file = file
+
+    @property
+    @functools.cache
+    def name(self):
+        """Basename from self.file without file extension and
+        with dashes replaced by underscores"""
+        return os.path.splitext(os.path.basename(self.file))[0].replace("-", "_")
+
+    @staticmethod
+    def _convert_key(name):
+        """Convert from camelCase to snake_case"""
+        return "_".join(Generator.SPLITTER.findall(name)).lower()
+
+    @staticmethod
+    def _convert_text(text):
+        """Replace gecko's placeholders with Qt's and
+        remove access key hints"""
+        def replacer(match):
+            return "%1" if match.group(1) is None else "%" + match.group(1)
+        new = Generator.PLACEHOLDER.sub(replacer, text)
+        while new != text:
+            text = new
+            new = Generator.PLACEHOLDER.sub(replacer, text)
+        return new.replace("&", "")
+
+    def process_line(self, line):
+        match = self.RE.match(line)
+        if match is not None:
+            return Translation(self.name, match.group('name'),
+                               self._convert_key(match.group('name')),
+                               self._convert_text(match.group('text')))
+
+    def __iter__(self):
+        try:
+            with open(self.file) as file:
+                for line in file:
+                    result = self.process_line(line)
+                    if result is not None:
+                        yield result
+        except UnicodeDecodeError:
+            print(f"Invalid character in file: {self.file}", file=sys.stderr)
+            raise
+
+@dataclasses.dataclass
+class Translation:
+    name: str
+    original_key: str
+    key: str
+    text: str
+
+    prefix = None
+
+    def __str__(self):
+        return f"""
+//: The original translation key - {self.original_key}
```

Comment:
Small tweak for this would be in place...

//: This is transformed translation key - the original translation key is {self.original_key} from file {self.source}

If source could be passed separately then script user could indicate nicely for example https://git.sailfishos.org/mer-core/gecko-dev/blob/master/toolkit/locales/en-US/chrome/passwordmgr/passwordmgr.properties

Reply:
I wonder if it would be useful to use urls instead of file names and then download the sources to build these files.

Reply:
It wasn't quite so small tweak but now it's possible to have those urls as well. For example:
generate_property_translations.py --url https://git.sailfishos.org/mer-core/gecko-dev/blob/master/ toolkit/locales/en-US/chrome/passwordmgr/passwordmgr.properties toolkit/locales/en-US/chrome/global/filepicker.properties
generates proper urls to the files but it means that this must be run on top level in that repository. Also version can be defined similarly.

Reply:
Seeing something like this as an output without version that is:

```
//: This is a transformed translation key,
//: the original translation key (insecureFieldWarningDescription3) is from https://git.sailfishos.org/mer-core/gecko-dev/blob/master/toolkit/locales/en-US/chrome/passwordmgr/passwordmgr.properties../gecko-dev/toolkit/locales/en-US/chrome/passwordmgr/passwordmgr.properties (unknown)
```

As I already gave the url I'd not expect to see that local file path. Ran like:

```
./tools/generate_property_translations.py  --url https://git.sailfishos.org/mer-core/gecko-dev/blob/master/toolkit/locales/en-US/chrome/passwordmgr/passwordmgr.properties ../gecko-dev/toolkit/locales/en-US/chrome/passwordmgr/passwordmgr.properties
```

Reply:
Well, the idea was that the translation file could be generated from more than one property file, thus the argument for --url needs to be common for all files. If that's not needed I can drop that feature and change it to be used like you described.

Reply:
Changed this as discussed yesterday. Removed --version argument and support for multiple input files. Also translation ids no longer have file name in them.